src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerPaper.java

```diff
@@ -0,0 +1,253 @@
+		if (localPlayer != null && localPlayer.isOnline()) {
+			refreshLocalPlayer(localPlayer);
+			return true;
+		}
+		return false;
+	}
+
+	// Run this on local players whenever their information is out of date
+	boolean refreshLocalPlayer(Player player) {
+		MMLog.fine(() -> "Refreshing local player " + player.getName());
+		RemotePlayerMinecraft localPlayer = fromLocal(player, true);
+
+		// update local player with data'
+		if (updateLocalPlayer(localPlayer, false)) {
+			localPlayer.broadcast();
+		}
+		return true;
+	}
+
+	// We recieved data from another server, add more data
+	void remotePlayerChange(JsonObject data) {
+		if (data == null) {
+			MMLog.severe(() -> "Null player data recieved from an unknown source!");
+			return;
+		}
+		RemotePlayerAbstraction player = RemotePlayerAbstraction.from(data);
+		if (player == null) {
+			MMLog.severe(() -> "Invalid player data recieved from an unknown source!");
+			return;
+		}
+
+		updateLocalPlayer(player, true);
+	}
+
+	@Override
+	void callPlayerLoadEvent(RemotePlayerAbstraction player) {
+		RemotePlayerLoadedEvent remotePE = new RemotePlayerLoadedEvent(player);
+		Bukkit.getServer().getPluginManager().callEvent(remotePE);
+	}
+
+	@Override
+	void callPlayerUnloadEvent(RemotePlayerAbstraction player) {
+		RemotePlayerUnloadedEvent remotePE = new RemotePlayerUnloadedEvent(player);
+		Bukkit.getServer().getPluginManager().callEvent(remotePE);
+	}
+
+	@Override
+	void callPlayerUpdatedEvent(RemotePlayerAbstraction player) {
+		RemotePlayerUpdatedEvent remotePE = new RemotePlayerUpdatedEvent(player);
+		Bukkit.getServer().getPluginManager().callEvent(remotePE);
+	}
+
+	@Override
+	Map<String, JsonObject> callGatherPluginDataEvent(RemotePlayerAbstraction player) {
+		GatherRemotePlayerDataEvent remotePE = new GatherRemotePlayerDataEvent(player);
+		Bukkit.getServer().getPluginManager().callEvent(remotePE);
+		return remotePE.getPluginData();
+	}
+
+	@Override
+	boolean checkAndRefreshIfLocalPlayer(RemotePlayerAbstraction player) {
+		if (!player.getServerType().equals(RemotePlayerMinecraft.SERVER_TYPE)) {
+			return false;
+		}
+		return refreshLocalPlayer(player.mUuid);
+	}
+
+	@EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
+	public void destOnlineEvent(DestOnlineEvent event) {
+		String remoteShardName = event.getDest();
+		if (getServerId().equals(remoteShardName)) {
+			return;
+		}
+		registerServer(remoteShardName);
+	}
+
+	@EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+	public void destOfflineEvent(DestOfflineEvent event) {
+		String remoteShardName = event.getDest();
+		unregisterServer(remoteShardName);
+	}
+
+	// Player ran a command
+	@EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
+	public void playerCommandPreprocessEvent(PlayerCommandPreprocessEvent event) {
+		String command = event.getMessage();
+
+		if (command.startsWith("/pv ")
+			|| "/pv".equals(command)
+			|| command.contains("vanish")) {
```

Comment:
These checks seem a bit weird - trying to be careful with pv but vanish could appear anywhere. Fortunately doesn't matter much since the worst case scenario is refreshing slightly more than necessary. Maybe could get the substring of the command ending until the first space and check that that is exactly "/pv" or "vanish"?

Reply:
Yeah, this is probably overly cautious, though it's the same check as in the chat plugin now. There just isn't really a good way to check if a player's vanished state changes without using `/pv` with optional arguments, or hoping that if another vanish command is in use, it contains `vanish`. Though I should probably at least include a leading `/` in all cases so it can't trigger on general chat.